### PR TITLE
Add test to make sure the new registry pass-thru token is only sent to the intended hosts

### PR DESCRIPTION
**Pull request by tiborvass**
Signed-off-by: Tibor Vass tibor@docker.com

Ping @stevvooe @pdevine @dmcgowan @dhiltgen

### CODE REVIEW

**Comment by dmp42** [comment]
cc @docker/distribution-maintainers

**Comment by vdemeester** [comment]
@tiborvass 
```
02:48:52 time="2016-01-26T02:48:52Z" level=error msg="image verification failed for digest sha256:02fee8c3220ba806531f606525eceb83f4feb654f62b207191b1c9209188dedd" 
02:48:52 --- FAIL: TestTokenPassThru (0.01s)
02:48:52 	registry_unit_test.go:84: Failed to receive registry token
02:48:52 FAIL
```

**Comment by tiborvass** [comment]
@vdemeester yeah i know, i'm trying to get the test to work. I changed `httptest.NewServer` to `httptest.NewTLSServer` but the client request doesn't know the certs.

**Comment by dmcgowan** [comment]
LGTM

File: distribution/registry_unit_test.go [hunk 1]
```diff
@@ -79,9 +83,39 @@ func TestTokenPassThru(t *testing.T) {
...
 	}
+}
+
+func TestTokenPassThruDifferentHost(t *testing.T) {
+	handler := new(tokenPassThruHandler)
+	ts := httptest.NewServer(handler)
+	defer ts.Close()
+
+	tsproxy := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		http.Redirect(w, r, ts.URL, http.StatusMovedPermanently)
```
Comment (tiborvass) [inline distribution/registry_unit_test.go]: this needs to be `ts.URL + "/v2/"`

**Comment by tiborvass** [comment]
Ping @dmcgowan @stevvooe I reopened this PR and fixed the test. The issue was that I should have sent the 401 from the redirecthandler.

**Comment by stevvooe** [comment]
@tiborvass So, is the assertion that we are not leaking this token because the redirect target does not emit a 401? What if the target emits a 401? Doesn't that then register the token with the challenge manner and include the token on subsequent requests?

**Comment by tiborvass** [comment]
@stevvooe fixed

**Comment by stevvooe** [comment]
LGTM

@tiborvass It seems like we can't reproduce the scenario I'm describing, so we'll have to go forward. I'll set aside some time to better understand this in the future. Thanks for taking a deep look into this.

**Comment by dmcgowan** [comment]
LGTM

> So, is the assertion that we are not leaking this token because the redirect target does not emit a 401? What if the target emits a 401? Doesn't that then register the token with the challenge manner and include the token on subsequent requests?

@stevvooe If the initial response which produces the challenges redirects then 401s (such as registry-1.somewhere.com/v2/ -> registry.somewhere.com/v2/), then the authentication will only get added for the redirected location (registry.somewhere.com/v2/) since that is where the challenge came from. Since the redirect logic already correctly handles these sort of redirects, it should not be an issue as it would be expected that all requests have the same redirection. I understand this is kind of difficult to explain a short response, lmk if you want to discuss further.

**Comment by stevvooe** [comment]
@dmcgowan I'm not sure if I have the cognitive load to better understand this. It seems to be that if we use the same token, and if the redirect target issues a 401, it would leak. Following from that, we are relying on the redirect target not to issue a 401 to avoid leaking the token. Whether this is reproducible in practice is unclear.

**Pull request merged**